Segmentation violation in cluster mode and latest docker container

Hi guys!
I've just tested latest version of NATS Streaming in cluster mode with 3 nodes and got `panic: runtime error: invalid memory address or nil pointer dereference`.

My steps:
1) Create 3 data directories (`srv1`, `srv2`, `srv3`)
2) Put simple `nats.conf` in each data directory (ports end with 1, 2 and 3 respectively):
```
port: 4231
http_port: 8231
store: "file"
dir: "/data"
cluster {
  host: '10.10.0.100'
  port: 5231
  bootstrap: true
  raft_logging: true
  routes = [
    nats-route://10.10.0.100:5232
    nats-route://10.10.0.100:5233
  ]
}
```
3) Start 3 docker containers:
`docker container run -it --name nats1 -v ${PWD}/srv1:/data --network=host nats-streaming -c /data/nats.conf`
4) Do some simple tests to be sure everything is working.
5) Start 3 consumers (just commented out printing every messages):
```
stan-sub -s nats://10.10.0.100:4231 -id 1 --qgroup qgtest -all test
stan-sub -s nats://10.10.0.100:4232 -id 2 --qgroup qgtest -all test
stan-sub -s nats://10.10.0.100:4233 -id 3 --qgroup qgtest -all test
```
6) Start data producer:
```
stan-bench -s nats://10.10.0.100:4231 -np 1 -ns 0 -n 10000000 -ms 1000 -a test
```

After several minutes second and third containers started printing errors:
```
...
[1] 2018/02/14 20:32:31.100452 [ERR] STREAM: [Client:benchmark-pub-1] Error replicating message for subject "_STAN.pub.test-cluster.test": timed out enqueuing operation
[1] 2018/02/14 20:32:31.100464 [ERR] STREAM: [Client:benchmark-pub-97] Error replicating message for subject "_STAN.pub.test-cluster.test": timed out enqueuing operation
[1] 2018/02/14 20:32:31.100475 [ERR] STREAM: [Client:benchmark-pub-94] Error replicating message for subject "_STAN.pub.test-cluster.test": timed out enqueuing operation
[1] 2018/02/14 20:32:31.100485 [ERR] STREAM: [Client:benchmark-pub-97] Error replicating message for subject "_STAN.pub.test-cluster.test": timed out enqueuing operation
...
```

and then:
```
[1] 2018/02/14 20:32:27.598057 [ERR] STREAM: [Client:benchmark-pub-85] Error replicating message for subject "_STAN.pub.test-cluster.test": timed out enqueuing operation
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x87e359]

goroutine 70 [running]:
github.com/nats-io/nats-streaming-server/stores.(*FileMsgStore).removeFirstMsg(0xc42009a340, 0x0, 0xc42ec9af00)
	/go/src/github.com/nats-io/nats-streaming-server/stores/filestore.go:2647 +0x69
github.com/nats-io/nats-streaming-server/stores.(*FileMsgStore).enforceLimits(0xc42009a340, 0x1e0001, 0xc4531b12c0, 0xc447f7b801)
	/go/src/github.com/nats-io/nats-streaming-server/stores/filestore.go:2589 +0xa2
github.com/nats-io/nats-streaming-server/stores.(*FileMsgStore).Store(0xc42009a340, 0xc4531b12c0, 0x0, 0x0, 0x0)
	/go/src/github.com/nats-io/nats-streaming-server/stores/filestore.go:2482 +0xb99
github.com/nats-io/nats-streaming-server/server.(*StanServer).Apply(0xc4201ea000, 0xc450f849a0, 0xdfe320, 0x3)
	/go/src/github.com/nats-io/nats-streaming-server/server/server.go:4498 +0x176
github.com/nats-io/nats-streaming-server/vendor/github.com/hashicorp/raft.(*Raft).runFSM.func1(0xc44b13b1e0)
	/go/src/github.com/nats-io/nats-streaming-server/vendor/github.com/hashicorp/raft/fsm.go:57 +0x17b
github.com/nats-io/nats-streaming-server/vendor/github.com/hashicorp/raft.(*Raft).runFSM(0xc4200e1340)
	/go/src/github.com/nats-io/nats-streaming-server/vendor/github.com/hashicorp/raft/fsm.go:120 +0x31e
github.com/nats-io/nats-streaming-server/vendor/github.com/hashicorp/raft.(*Raft).(github.com/nats-io/nats-streaming-server/vendor/github.com/hashicorp/raft.runFSM)-fm()
	/go/src/github.com/nats-io/nats-streaming-server/vendor/github.com/hashicorp/raft/api.go:506 +0x2a
github.com/nats-io/nats-streaming-server/vendor/github.com/hashicorp/raft.(*raftState).goFunc.func1(0xc4200e1340, 0xc420180500)
	/go/src/github.com/nats-io/nats-streaming-server/vendor/github.com/hashicorp/raft/state.go:146 +0x53
created by github.com/nats-io/nats-streaming-server/vendor/github.com/hashicorp/raft.(*raftState).goFunc
	/go/src/github.com/nats-io/nats-streaming-server/vendor/github.com/hashicorp/raft/state.go:144 +0x66
```

After restarting failed containers:
```
$ docker container start -ai nats2
[1] 2018/02/14 20:34:30.879042 [INF] STREAM: Starting nats-streaming-server[test-cluster] version 0.8.0-beta
[1] 2018/02/14 20:34:30.879082 [INF] STREAM: ServerID: PIWf5lutLWcVp9fYVr2WFY
[1] 2018/02/14 20:34:30.879087 [INF] STREAM: Go version: go1.9.3
[1] 2018/02/14 20:34:30.879323 [INF] Starting nats-server version 1.0.4
[1] 2018/02/14 20:34:30.879554 [INF] Starting http monitor on 0.0.0.0:8232
[1] 2018/02/14 20:34:30.879671 [INF] Listening for client connections on 0.0.0.0:4232
[1] 2018/02/14 20:34:30.879685 [INF] Server is ready
[1] 2018/02/14 20:34:30.880552 [INF] Listening for route connections on 10.10.0.100:5232
[1] 2018/02/14 20:34:30.906597 [INF] STREAM: Recovering the state...
[1] 2018/02/14 20:34:31.373113 [INF] STREAM: Recovered 1 channel(s)
[1] 2018/02/14 20:34:33.298825 [INF] STREAM: raft: Initial configuration (index=1): [{Suffrage:Voter ID:ib2LAFTTccvyusMmcYgM9D Address:test-cluster.ib2LAFTTccvyusMmcYgM9D.stan}]
[1] 2018/02/14 20:34:33.298939 [INF] STREAM: raft: Node at test-cluster.ib2LAFTTccvyusMmcYgM9D.stan [Follower] entering Follower state (Leader: "")
[1] 2018/02/14 20:34:33.300337 [INF] STREAM: Message store is RAFT_FILE
[1] 2018/02/14 20:34:33.300415 [INF] STREAM: ---------- Store Limits ----------
[1] 2018/02/14 20:34:33.300425 [INF] STREAM: Channels:                  100 *
[1] 2018/02/14 20:34:33.300431 [INF] STREAM: --------- Channels Limits --------
[1] 2018/02/14 20:34:33.300436 [INF] STREAM:   Subscriptions:          1000 *
[1] 2018/02/14 20:34:33.300443 [INF] STREAM:   Messages     :       1000000 *
[1] 2018/02/14 20:34:33.300449 [INF] STREAM:   Bytes        :     976.56 MB *
[1] 2018/02/14 20:34:33.300456 [INF] STREAM:   Age          :     unlimited *
[1] 2018/02/14 20:34:33.300464 [INF] STREAM: ----------------------------------
[1] 2018/02/14 20:34:35.169644 [INF] STREAM: raft: Heartbeat timeout from "" reached, starting election
[1] 2018/02/14 20:34:35.169694 [INF] STREAM: raft: Node at test-cluster.ib2LAFTTccvyusMmcYgM9D.stan [Candidate] entering Candidate state in term 3
[1] 2018/02/14 20:34:35.169961 [INF] STREAM: raft: Election won. Tally: 1
[1] 2018/02/14 20:34:35.169977 [INF] STREAM: raft: Node at test-cluster.ib2LAFTTccvyusMmcYgM9D.stan [Leader] entering Leader state
[1] 2018/02/14 20:34:35.170079 [INF] STREAM: server became leader, performing leader promotion actions
[1] 2018/02/14 20:34:38.276052 [INF] STREAM: WARNING: Reached limits for store "test" (msgs=963311/1000000 bytes=976.56 MB/976.56 MB), dropping old messages to make room for new ones
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x87e359]

goroutine 92 [running]:
github.com/nats-io/nats-streaming-server/stores.(*FileMsgStore).removeFirstMsg(0xc42023a000, 0x0, 0xc4203e1600)
	/go/src/github.com/nats-io/nats-streaming-server/stores/filestore.go:2647 +0x69
github.com/nats-io/nats-streaming-server/stores.(*FileMsgStore).enforceLimits(0xc42023a000, 0x1e0001, 0xc42172aa50, 0xc422016a01)
	/go/src/github.com/nats-io/nats-streaming-server/stores/filestore.go:2589 +0xa2
github.com/nats-io/nats-streaming-server/stores.(*FileMsgStore).Store(0xc42023a000, 0xc42172aa50, 0x0, 0x0, 0x0)
	/go/src/github.com/nats-io/nats-streaming-server/stores/filestore.go:2482 +0xb99
github.com/nats-io/nats-streaming-server/server.(*StanServer).Apply(0xc4201ee000, 0xc4203354a0, 0xdfe320, 0xbe99453790618e89)
	/go/src/github.com/nats-io/nats-streaming-server/server/server.go:4498 +0x176
github.com/nats-io/nats-streaming-server/vendor/github.com/hashicorp/raft.(*Raft).runFSM.func1(0xc4202ec2d0)
	/go/src/github.com/nats-io/nats-streaming-server/vendor/github.com/hashicorp/raft/fsm.go:57 +0x17b
github.com/nats-io/nats-streaming-server/vendor/github.com/hashicorp/raft.(*Raft).runFSM(0xc4200d0000)
	/go/src/github.com/nats-io/nats-streaming-server/vendor/github.com/hashicorp/raft/fsm.go:120 +0x31e
github.com/nats-io/nats-streaming-server/vendor/github.com/hashicorp/raft.(*Raft).(github.com/nats-io/nats-streaming-server/vendor/github.com/hashicorp/raft.runFSM)-fm()
	/go/src/github.com/nats-io/nats-streaming-server/vendor/github.com/hashicorp/raft/api.go:506 +0x2a
github.com/nats-io/nats-streaming-server/vendor/github.com/hashicorp/raft.(*raftState).goFunc.func1(0xc4200d0000, 0xc4202ec150)
	/go/src/github.com/nats-io/nats-streaming-server/vendor/github.com/hashicorp/raft/state.go:146 +0x53
created by github.com/nats-io/nats-streaming-server/vendor/github.com/hashicorp/raft.(*raftState).goFunc
	/go/src/github.com/nats-io/nats-streaming-server/vendor/github.com/hashicorp/raft/state.go:144 +0x66
```

--- a/server/clustering_test.go
+++ b/server/clustering_test.go
@@ -16,6 +16,7 @@ import (
 	"testing"
 	"time"
 
+	"github.com/hashicorp/raft"
 	natsdTest "github.com/nats-io/gnatsd/test"
 	"github.com/nats-io/go-nats"
 	"github.com/nats-io/go-nats-streaming"
@@ -628,7 +629,7 @@ func TestClusteringNoPanicOnShutdown(t *testing.T) {
 	// Wait for leader to be elected.
 	leader := getLeader(t, 10*time.Second, servers...)
 
-	sc, err := stan.Connect(clusterName, clientName, stan.PubAckWait(time.Second), stan.ConnectWait(10*time.Second))
+	sc, err := stan.Connect(clusterName, clientName, stan.PubAckWait(time.Second))
 	if err != nil {
 		t.Fatalf("Error on connect: %v", err)
 	}
@@ -658,12 +659,21 @@ func TestClusteringNoPanicOnShutdown(t *testing.T) {
 	time.Sleep(time.Duration(rand.Intn(500)+100) * time.Millisecond)
 
 	// We shutdown the follower, it should not panic.
+	follower := s1
 	if s1 == leader {
-		s2.Shutdown()
-	} else {
-		s1.Shutdown()
+		follower = s2
 	}
+	follower.Shutdown()
 	wg.Wait()
+
+	// Restart follower to speed up client disconnect.
+	follower = runServerWithOpts(t, follower.opts, nil)
+	defer follower.Shutdown()
+
+	// Close client explicitly otherwise follower will be shutdown
+	// first and then client close will timeout since only 1 node
+	// in cluster (no leader)
+	sc.Close()
 }
 
 func TestClusteringLeaderFlap(t *testing.T) {
@@ -2488,4 +2498,255 @@ func TestClusteringAndChannelsPartitioning(t *testing.T) {
 		s.Shutdown()
 		t.Fatal("Expected error, got none")
 	}
+}
+
+func TestClusteringGetProperErrorFromFSMApply(t *testing.T) {
+	cleanupDatastore(t)
+	defer cleanupDatastore(t)
+	cleanupRaftLog(t)
+	defer cleanupRaftLog(t)
+
+	// For this test, use a central NATS server.
+	ns := natsdTest.RunDefaultServer()
+	defer ns.Shutdown()
+
+	// Configure first server
+	s1sOpts := getTestDefaultOptsForClustering("a", true)
+	s1 := runServerWithOpts(t, s1sOpts, nil)
+	defer s1.Shutdown()
+
+	// Configure second server.
+	s2sOpts := getTestDefaultOptsForClustering("b", false)
+	s2 := runServerWithOpts(t, s2sOpts, nil)
+	defer s2.Shutdown()
+
+	servers := []*StanServer{s1, s2}
+
+	// Wait for leader to be elected.
+	leader := getLeader(t, 10*time.Second, servers...)
+
+	sc1, err := stan.Connect(clusterName, clientName)
+	if err != nil {
+		t.Fatalf("Error on connect: %v", err)
+	}
+	defer sc1.Close()
+
+	if _, err := sc1.Subscribe("foo", func(_ *stan.Msg) {}, stan.DurableName("dur")); err != nil {
+		t.Fatalf("Error on subscribe: %v", err)
+	}
+	// This one should fail
+	if _, err := sc1.Subscribe("foo", func(_ *stan.Msg) {}, stan.DurableName("dur")); err == nil {
+		t.Fatal("Creation of durable should have failed")
+	}
+
+	// We will send crafted unsubscribe protocol with wrong clientID
+	waitForNumSubs(t, leader, clientName, 1)
+	// Get the sub, so we can get required info for unsub request
+	sub := leader.clients.getSubs(clientName)[0]
+	unsubReq := &pb.UnsubscribeRequest{
+		Inbox:       sub.Inbox,
+		Subject:     sub.subject,
+		DurableName: sub.DurableName,
+		ClientID:    "wrongcid",
+	}
+	data, _ := unsubReq.Marshal()
+	reply, err := sc1.NatsConn().Request(leader.info.Unsubscribe, data, 2*time.Second)
+	if err != nil {
+		t.Fatalf("Error on request: %v", err)
+	}
+	unsubReply := &pb.SubscriptionResponse{}
+	unsubReply.Unmarshal(reply.Data)
+	if unsubReply.Error == "" {
+		t.Fatal("Unsubscribe should have returned an error")
+	}
+
+	leader.clients.Lock()
+	orgClientStore := leader.clients.store
+	leader.clients.store = &clientStoreErrorsStore{Store: orgClientStore}
+	leader.clients.Unlock()
+
+	// These operations should fail
+
+	// Server should fail to store client
+	sc2, err := stan.Connect(clusterName, clientName+"2")
+	if err == nil {
+		sc2.Close()
+		t.Fatal("Second connect should have failed")
+	}
+	// Server should fail to delete this unknown client
+	closeReq := &pb.CloseRequest{ClientID: "wrongid"}
+	data, _ = closeReq.Marshal()
+	reply, err = sc1.NatsConn().Request(leader.info.Close, data, 2*time.Second)
+	if err != nil {
+		t.Fatalf("Error on request: %v", err)
+	}
+	closeReply := &pb.CloseResponse{}
+	closeReply.Unmarshal(reply.Data)
+	if closeReply.Error == "" {
+		t.Fatal("Close should have returned an error")
+	}
+	// Restore client store to speed up end of test
+	leader.clients.Lock()
+	leader.clients.store = orgClientStore
+	leader.clients.Unlock()
+}
+
+func TestClusteringNoMsgSeqGapOnApplyError(t *testing.T) {
+	cleanupDatastore(t)
+	defer cleanupDatastore(t)
+	cleanupRaftLog(t)
+	defer cleanupRaftLog(t)
+
+	// For this test, use a central NATS server.
+	ns := natsdTest.RunDefaultServer()
+	defer ns.Shutdown()
+
+	// Configure first server
+	s1sOpts := getTestDefaultOptsForClustering("a", true)
+	s1 := runServerWithOpts(t, s1sOpts, nil)
+	defer s1.Shutdown()
+
+	// Configure second server.
+	s2sOpts := getTestDefaultOptsForClustering("b", false)
+	s2 := runServerWithOpts(t, s2sOpts, nil)
+	defer s2.Shutdown()
+
+	servers := []*StanServer{s1, s2}
+
+	// Wait for leader to be elected.
+	leader := getLeader(t, 10*time.Second, servers...)
+
+	sc, err := stan.Connect(clusterName, clientName,
+		stan.MaxPubAcksInflight(100),
+		stan.PubAckWait(500*time.Millisecond))
+	if err != nil {
+		t.Fatalf("Error on connect: %v", err)
+	}
+	defer sc.Close()
+
+	follower := s2
+	if s2 == leader {
+		follower = s1
+	}
+
+	stopCh := make(chan struct{}, 1)
+	errCount := uint32(0)
+	errLeadeshipLost := uint32(0)
+	okCount := uint32(0)
+	sent := uint32(0)
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		i := 1
+		for {
+			select {
+			case <-stopCh:
+				return
+			default:
+				if _, err := sc.PublishAsync("foo", []byte(fmt.Sprintf("%d", i)), func(guid string, err error) {
+					if err != nil {
+						atomic.AddUint32(&errCount, 1)
+						if err.Error() == raft.ErrLeadershipLost.Error() {
+							atomic.AddUint32(&errLeadeshipLost, 1)
+						}
+					} else {
+						atomic.AddUint32(&okCount, 1)
+					}
+				}); err != nil {
+					return
+				}
+				atomic.AddUint32(&sent, 1)
+				i++
+			}
+		}
+	}()
+	// Wait so that go-routine is in middle of sending messages
+	time.Sleep(time.Duration(rand.Intn(500)+100) * time.Millisecond)
+	// Shutdow follower
+	follower.Shutdown()
+	// The leader should step down since there is only 1 node
+	verifyNoLeader(t, 2*time.Second, servers...)
+	// stop the publisher loop
+	stopCh <- struct{}{}
+	wg.Wait()
+
+	timeout := time.Now().Add(5 * time.Second)
+	ok := false
+	for time.Now().Before(timeout) {
+		total := atomic.LoadUint32(&okCount) + atomic.LoadUint32(&errCount)
+		if total == atomic.LoadUint32(&sent) {
+			ok = true
+			break
+		}
+		time.Sleep(15 * time.Millisecond)
+	}
+	if !ok {
+		t.Fatalf("Timedout waiting for total sent messages")
+	}
+
+	okInStore := uint64(atomic.LoadUint32(&okCount))
+
+	ch := make(chan struct{}, 1)
+	leader.ioChannel <- &ioPendingMsg{sc: ch}
+	<-ch
+	// Get the old leader's store first/last sequence
+	c := leader.channels.get("foo")
+	first, last := msgStoreFirstAndLastSequence(t, c.store.Msgs)
+	if first != 1 || last != okInStore {
+		t.Fatalf("Expected first/last to be %v/%v got %v/%v", 1, okInStore, first, last)
+	}
+	if c.nextSequence != last+1 {
+		t.Fatalf("Expected channel nextSequence to be %v, got %v", last+1, c.nextSequence)
+	}
+	close(ch)
+
+	// Restart the follower
+	servers = removeServer(servers, follower)
+	follower = runServerWithOpts(t, follower.opts, nil)
+	defer follower.Shutdown()
+	servers = append(servers, follower)
+
+	leader = getLeader(t, 10*time.Second, servers...)
+
+	// Publish last message
+	if err := sc.Publish("foo", []byte("last")); err != nil {
+		t.Fatalf("Error on publish: %v", err)
+	}
+
+	leader.raft.Barrier(0).Error()
+
+	okInStore = uint64(atomic.LoadUint32(&okCount) + atomic.LoadUint32(&errLeadeshipLost) + 1)
+	c = leader.channels.get("foo")
+	timeout = time.Now().Add(10 * time.Second)
+	ok = false
+	for time.Now().Before(timeout) {
+		first, last = msgStoreFirstAndLastSequence(t, c.store.Msgs)
+		if first == 1 && last == okInStore {
+			ok = true
+			break
+		}
+		time.Sleep(15 * time.Millisecond)
+	}
+	if !ok {
+		t.Fatalf("Timedout wait for all messages to be stored... expected %v, got %v", okInStore, last)
+	}
+	for i := first; i <= last; i++ {
+		m, err := c.store.Msgs.Lookup(i)
+		if err != nil {
+			t.Fatalf("Error on lookup for seq=%v: %v", i, err)
+		}
+		if m == nil {
+			t.Fatalf("Nil message for seq: %v", i)
+		}
+		if i != last {
+			mi, _ := strconv.Atoi(string(m.Data))
+			if i != uint64(mi) {
+				t.Fatalf("Expected seq %v, got %v (m=%v)", i, mi, m)
+			}
+		} else if string(m.Data) != "last" {
+			t.Fatalf("Unexpected last message: %v", m)
+		}
+	}
+	sc.Close()
 }

--- a/server/server.go
+++ b/server/server.go
@@ -97,9 +97,6 @@ const (
 	// Name of the file to store Raft log.
 	raftLogFile = "raft.log"
 
-	// Time to wait on starting a Raft operation.
-	raftApplyTimeout = 5 * time.Second
-
 	// In partitioning mode, when a client connects, the connect request
 	// may reach several servers, but the first response the client gets
 	// allows it to proceed with either publish or subscribe.
@@ -186,6 +183,7 @@ type ioPendingMsg struct {
 	pm pb.PubMsg
 	pa pb.PubAck
 	c  *channel
+	sc chan struct{}
 }
 
 // Constant that defines the size of the channel that feeds the IO thread.
@@ -304,7 +302,7 @@ func (cs *channelStore) create(s *StanServer, name string, sc *stores.Channel) (
 	if err != nil {
 		return nil, err
 	}
-	atomic.StoreUint64(&c.nextSequence, lastSequence+1)
+	c.nextSequence = lastSequence + 1
 	cs.channels[name] = c
 	return c, nil
 }
@@ -1234,7 +1232,7 @@ func RunServerWithOpts(stanOpts *Options, natsOpts *server.Options) (newServer *
 		opts:          sOpts,
 		natsOpts:      nOpts,
 		dupCIDTimeout: defaultCheckDupCIDTimeout,
-		ioChannelQuit: make(chan struct{}, 1),
+		ioChannelQuit: make(chan struct{}),
 		trace:         sOpts.Trace,
 		debug:         sOpts.Debug,
 		subStartCh:    make(chan *subStartInfo, defaultSubStartChanLen),
@@ -1660,6 +1658,23 @@ func (s *StanServer) leadershipAcquired() error {
 	s.log.Noticef("server became leader, performing leader promotion actions")
 	defer s.log.Noticef("finished leader promotion actions")
 
+	// If we were not the leader, there should be nothing in the ioChannel
+	// (processing of client publishes). However, since a node could go
+	// from leader to follower to leader again, let's make sure that we
+	// synchronize with the ioLoop before we touch the channels' nextSequence.
+	ch := make(chan struct{}, 1)
+	iopm := &ioPendingMsg{sc: ch}
+	s.ioChannel <- iopm
+	// Wait for the ioLoop to reach that special iopm and notifies us (or
+	// give up if server is shutting down).
+	select {
+	case <-ch:
+	case <-s.ioChannelQuit:
+		return nil
+	}
+	// Then, we will notify it back to unlock it when were are done here.
+	defer close(ch)
+
 	// Use a barrier to ensure all preceding operations are applied to the FSM
 	if err := s.raft.Barrier(0).Error(); err != nil {
 		return err
@@ -1672,7 +1687,7 @@ func (s *StanServer) leadershipAcquired() error {
 		if err != nil {
 			return err
 		}
-		atomic.StoreUint64(&c.nextSequence, lastSequence+1)
+		c.nextSequence = lastSequence + 1
 	}
 
 	// Setup client heartbeats.
@@ -2302,6 +2317,25 @@ func (s *StanServer) isDuplicateConnect(client *client) bool {
 	return err == nil
 }
 
+// When calling ApplyFuture.Error(), if we get an error it means
+// that raft failed to commit to its log.
+// But if we also want the result of FSM.Apply(), which in this
+// case is StanServer.Apply(), we need to check the Response().
+// So we first check error from future.Error(). If nil, then we
+// check the Response.
+func waitForReplicationErrResponse(f raft.ApplyFuture) error {
+	err := f.Error()
+	if err == nil {
+		resp := f.Response()
+		// We call this function when we know that FutureApply's
+		// Response is an error object.
+		if resp != nil {
+			err = f.Response().(error)
+		}
+	}
+	return err
+}
+
 func (s *StanServer) replicateConnect(req *pb.ConnectRequest, refresh bool) error {
 	op := &spb.RaftOperation{
 		OpType: spb.RaftOperation_Connect,
@@ -2315,7 +2349,7 @@ func (s *StanServer) replicateConnect(req *pb.ConnectRequest, refresh bool) erro
 		panic(err)
 	}
 	// Wait on result of replication.
-	return s.raft.Apply(data, raftApplyTimeout).Error()
+	return waitForReplicationErrResponse(s.raft.Apply(data, 0))
 }
 
 func (s *StanServer) processConnect(req *pb.ConnectRequest, refresh bool) error {
@@ -2511,7 +2545,7 @@ func (s *StanServer) replicateConnClose(req *pb.CloseRequest) error {
 		panic(err)
 	}
 	// Wait on result of replication.
-	return s.raft.Apply(data, raftApplyTimeout).Error()
+	return waitForReplicationErrResponse(s.raft.Apply(data, 0))
 }
 
 func (s *StanServer) sendCloseResponse(subj string, closeErr error) {
@@ -2949,7 +2983,7 @@ func (s *StanServer) replicateSentAndAckSeqs(sub *subState) {
 	if err != nil {
 		panic(err)
 	}
-	s.raft.Apply(data, raftApplyTimeout)
+	s.raft.Apply(data, 0)
 	sub.replicate.reset()
 }
 
@@ -3176,16 +3210,6 @@ func (s *StanServer) startIOLoop() {
 func (s *StanServer) ioLoop(ready *sync.WaitGroup) {
 	defer s.ioChannelWG.Done()
 
-	////////////////////////////////////////////////////////////////////////////
-	// This is where we will store the message and wait for others in the
-	// potential cluster to do so as well, once we have a quorom someone can
-	// ack the publisher. We simply do so here for now.
-	////////////////////////////////////////////////////////////////////////////
-	////////////////////////////////////////////////////////////////////////////
-	// Once we have ack'd the publisher, we need to assign this a sequence ID.
-	// This will be done by a master election within the cluster, for now we
-	// assume we are the master and assign the sequence ID here.
-	////////////////////////////////////////////////////////////////////////////
 	storesToFlush := make(map[*channel]struct{}, 64)
 
 	var (
@@ -3193,42 +3217,74 @@ func (s *StanServer) ioLoop(ready *sync.WaitGroup) {
 		pendingMsgs  = _pendingMsgs[:0]
 	)
 
-	storeIOPendingMsgs := func(iopms []*ioPendingMsg) []raft.Future {
+	storeIOPendingMsgs := func(iopms []*ioPendingMsg) {
 		var (
-			futuresMap       map[*channel]raft.Future
-			replicateFutures []raft.Future
-			err              error
-			succeeded        []*ioPendingMsg
-			failed           []*ioPendingMsg
+			futuresMap map[*channel]raft.Future
+			err        error
 		)
 		if s.isClustered {
 			futuresMap, err = s.replicate(iopms)
+			// If replicate() returns an error, it means that no future
+			// was applied, so we can fail all published messages.
 			if err != nil {
-				failed = iopms
-			} else {
-				succeeded = iopms
-				replicateFutures = make([]raft.Future, len(iopms))
-
-				for c := range futuresMap {
-					storesToFlush[c] = struct{}{}
+				for _, iopm := range iopms {
+					s.logErrAndSendPublishErr(iopm, err)
 				}
-				for i, iopm := range iopms {
-					replicateFutures[i] = futuresMap[iopm.c]
+			} else {
+				for c, f := range futuresMap {
+					// Wait for the replication result.
+					// We know that we panic in StanServer.Apply() if storing
+					// of messages fail. So the only reason f.Error() would
+					// return an error (we are not using timeout in Apply())
+					// is if raft fails to store its log, but it would have
+					// then switched follower state. On leadership acquisition
+					// we do reet nextSequence based on lastSequence on store.
+					// Regardless, do reset here in case of error.
+
+					// Note that each future contains a batch of messages for
+					// a given channel. All futures in the map are for different
+					// channels.
+					if err := f.Error(); err != nil {
+						lastSeq, lerr := c.store.Msgs.LastSequence()
+						if lerr != nil {
+							panic(fmt.Errorf("Error during message replication (%v), unable to get store last sequence: %v", err, lerr))
+						}
+						c.nextSequence = lastSeq + 1
+					} else {
+						storesToFlush[c] = struct{}{}
+					}
+				}
+				// We have 1 future per channel. However, the array of iopms
+				// may be from different channels. For each iopm we look
+				// up its corresponding future and if there was an error
+				// (same for all iopms of the same channel) we fail the
+				// corresponding publishers.
+				for _, iopm := range iopms {
+					// We can call Error() again, this is not a problem.
+					if err := futuresMap[iopm.c].Error(); err != nil {
+						s.logErrAndSendPublishErr(iopm, err)
+					} else {
+						pendingMsgs = append(pendingMsgs, iopm)
+					}
 				}
 			}
 		} else {
-			succeeded, failed, err = s.assignAndStore(iopms, storesToFlush)
-		}
-		if err != nil {
-			for _, iopm := range failed {
-				s.log.Errorf("[Client:%s] Error processing message for subject %q: %v",
-					iopm.pm.ClientID, iopm.m.Subject, err)
-				s.sendPublishErr(iopm.m.Reply, iopm.pm.Guid, err)
+			for _, iopm := range iopms {
+				pm := &iopm.pm
+				c, err := s.lookupOrCreateChannel(pm.Subject)
+				if err == nil {
+					msg := c.pubMsgToMsgProto(pm, c.nextSequence)
+					_, err = c.store.Msgs.Store(msg)
+				}
+				if err != nil {
+					s.logErrAndSendPublishErr(iopm, err)
+				} else {
+					c.nextSequence++
+					pendingMsgs = append(pendingMsgs, iopm)
+					storesToFlush[c] = struct{}{}
+				}
 			}
 		}
-		pendingMsgs = append(pendingMsgs, succeeded...)
-
-		return replicateFutures
 	}
 
 	var (
@@ -3239,11 +3295,20 @@ func (s *StanServer) ioLoop(ready *sync.WaitGroup) {
 		batch     = make([]*ioPendingMsg, 0, batchSize)
 	)
 
+	leadershipAcquiredSync := func(iopm *ioPendingMsg) {
+		iopm.sc <- struct{}{}
+		<-iopm.sc
+	}
+
 	ready.Done()
 	for {
 		batch = batch[:0]
 		select {
 		case iopm := <-s.ioChannel:
+			if iopm.sc != nil {
+				leadershipAcquiredSync(iopm)
+				continue
+			}
 			batch = append(batch, iopm)
 
 			remaining := batchSize - 1
@@ -3272,7 +3337,12 @@ func (s *StanServer) ioLoop(ready *sync.WaitGroup) {
 				}
 
 				for i := 0; i < ioChanLen; i++ {
-					batch = append(batch, <-s.ioChannel)
+					iopm = <-s.ioChannel
+					if iopm.sc != nil {
+						leadershipAcquiredSync(iopm)
+					} else {
+						batch = append(batch, iopm)
+					}
 				}
 				// Keep track of max number of messages in a batch
 				if ioChanLen > max {
@@ -3282,20 +3352,8 @@ func (s *StanServer) ioLoop(ready *sync.WaitGroup) {
 				remaining -= ioChanLen
 			}
 
-			replicateFutures := storeIOPendingMsgs(batch)
-
 			// If clustered, wait on the result of replication.
-			if s.isClustered {
-				for i, future := range replicateFutures {
-					iopm = pendingMsgs[i]
-					if err := future.Error(); err != nil {
-						s.log.Errorf("[Client:%s] Error replicating message for subject %q: %v",
-							iopm.pm.ClientID, iopm.m.Subject, err)
-						s.sendPublishErr(iopm.m.Reply, iopm.pm.Guid, fmt.Errorf("replication error: %s", err))
-						pendingMsgs[i] = nil
-					}
-				}
-			}
+			storeIOPendingMsgs(batch)
 
 			// flush all the stores with messages written to them...
 			for c := range storesToFlush {
@@ -3316,10 +3374,6 @@ func (s *StanServer) ioLoop(ready *sync.WaitGroup) {
 			// Ack our messages back to the publisher
 			for i := range pendingMsgs {
 				iopm := pendingMsgs[i]
-				if iopm == nil {
-					// Was removed because replication failed.
-					continue
-				}
 				s.ackPublisher(iopm)
 				pendingMsgs[i] = nil
 			}
@@ -3333,23 +3387,10 @@ func (s *StanServer) ioLoop(ready *sync.WaitGroup) {
 	}
 }
 
-// assignAndStore will assign a sequence ID and then store the message. This
-// should not be called if running in clustered mode.
-func (s *StanServer) assignAndStore(iopms []*ioPendingMsg, storesToFlush map[*channel]struct{}) ([]*ioPendingMsg, []*ioPendingMsg, error) {
-	for i, iopm := range iopms {
-		pm := &iopm.pm
-		c, err := s.lookupOrCreateChannel(pm.Subject)
-		if err != nil {
-			return iopms[:i], iopms[i:], err
-		}
-		msg := c.pubMsgToMsgProto(pm, c.nextSequence)
-		if _, err := c.store.Msgs.Store(msg); err != nil {
-			return iopms[:i], iopms[i:], err
-		}
-		c.nextSequence++
-		storesToFlush[c] = struct{}{}
-	}
-	return iopms, nil, nil
+func (s *StanServer) logErrAndSendPublishErr(iopm *ioPendingMsg, err error) {
+	s.log.Errorf("[Client:%s] Error processing message for subject %q: %v",
+		iopm.pm.ClientID, iopm.m.Subject, err)
+	s.sendPublishErr(iopm.m.Reply, iopm.pm.Guid, err)
 }
 
 // replicate will replicate the batch of messages to followers and return
@@ -3367,7 +3408,7 @@ func (s *StanServer) replicate(iopms []*ioPendingMsg) (map[*channel]raft.Future,
 		if err != nil {
 			return nil, err
 		}
-		msg := c.pubMsgToMsgProto(pm, atomic.LoadUint64(&c.nextSequence))
+		msg := c.pubMsgToMsgProto(pm, c.nextSequence)
 		batch := batches[c]
 		if batch == nil {
 			batch = &spb.Batch{}
@@ -3375,7 +3416,7 @@ func (s *StanServer) replicate(iopms []*ioPendingMsg) (map[*channel]raft.Future,
 		}
 		batch.Messages = append(batch.Messages, msg)
 		iopm.c = c
-		atomic.AddUint64(&c.nextSequence, 1)
+		c.nextSequence++
 	}
 	for c, batch := range batches {
 		op := &spb.RaftOperation{
@@ -3386,7 +3427,7 @@ func (s *StanServer) replicate(iopms []*ioPendingMsg) (map[*channel]raft.Future,
 		if err != nil {
 			panic(err)
 		}
-		futures[c] = s.raft.Apply(data, raftApplyTimeout)
+		futures[c] = s.raft.Apply(data, 0)
 	}
 	return futures, nil
 }
@@ -3573,7 +3614,7 @@ func (s *StanServer) replicateUnsubscribe(req *pb.UnsubscribeRequest, opType spb
 		panic(err)
 	}
 	// Wait on result of replication.
-	return s.raft.Apply(data, raftApplyTimeout).Error()
+	return waitForReplicationErrResponse(s.raft.Apply(data, 0))
 }
 
 func (s *StanServer) sendSubscriptionResponseErr(reply string, err error) {
@@ -3687,12 +3728,12 @@ func (s *StanServer) replicateSub(sr *pb.SubscriptionRequest, ackInbox string) (
 		panic(err)
 	}
 	// Replicate operation and wait on result.
-	future := s.raft.Apply(data, raftApplyTimeout)
+	future := s.raft.Apply(data, 0)
 	if err := future.Error(); err != nil {
 		return nil, nil, err
 	}
 	rs := future.Response().(*replicatedSub)
-	return rs.c, rs.sub, nil
+	return rs.c, rs.sub, rs.err
 }
 
 // addSubscription adds `sub` to the client and store.
@@ -4402,7 +4443,7 @@ func (s *StanServer) Shutdown() {
 
 	if s.ioChannel != nil {
 		// Notify the IO channel that we are shutting down
-		s.ioChannelQuit <- struct{}{}
+		close(s.ioChannelQuit)
 	} else {
 		waitForIOStoreLoop = false
 	}
@@ -4460,6 +4501,7 @@ func (s *StanServer) Shutdown() {
 type replicatedSub struct {
 	c   *channel
 	sub *subState
+	err error
 }
 
 // Apply log is invoked once a log entry is committed.
@@ -4474,12 +4516,19 @@ func (s *StanServer) Apply(l *raft.Log) interface{} {
 	switch op.OpType {
 	case spb.RaftOperation_Publish:
 		// Message replication.
+		var (
+			c   *channel
+			err error
+		)
 		for _, msg := range op.PublishBatch.Messages {
-			c, err := s.lookupOrCreateChannel(msg.Subject)
-			if err != nil {
-				return err
+			// This is a batch for a given channel, so lookup channel once.
+			if c == nil {
+				c, err = s.lookupOrCreateChannel(msg.Subject)
 			}
-			if _, err := c.store.Msgs.Store(msg); err != nil {
+			if err == nil {
+				_, err = c.store.Msgs.Store(msg)
+			}
+			if err != nil {
 				panic(fmt.Errorf("failed to store replicated message %d on channel %s: %v",
 					msg.Sequence, c.name, err))
 			}
@@ -4494,10 +4543,7 @@ func (s *StanServer) Apply(l *raft.Log) interface{} {
 	case spb.RaftOperation_Subscribe:
 		// Subscription replication.
 		c, sub, err := s.processSub(op.Sub.Request, op.Sub.AckInbox)
-		if err != nil {
-			return err
-		}
-		return &replicatedSub{c: c, sub: sub}
+		return &replicatedSub{c: c, sub: sub, err: err}
 	case spb.RaftOperation_RemoveSubscription:
 		fallthrough
 	case spb.RaftOperation_CloseSubscription:

